testng-junit-migrator/src/test/java/tech/picnic/errorprone/testngjunit/TestNGScannerTest.java

```diff
@@ -0,0 +1,231 @@
+            "",
+            "  @BeforeMethod",
+            "  // BUG: Diagnostic contains: class: A setupMethod: BEFORE_METHOD",
+            "  private void beforeMethod() {}",
+            "",
+            "  @AfterClass",
+            "  // BUG: Diagnostic contains: class: A setupMethod: AFTER_CLASS",
+            "  private static void afterClass() {}",
+            "",
+            "  @AfterMethod",
+            "  // BUG: Diagnostic contains: class: A setupMethod: AFTER_METHOD",
+            "  private void afterMethod() {}",
+            "}")
+        .doTest();
+  }
+
+  /**
+   * A {@link BugChecker} that flags classes with a diagnostics message that indicates, whether a
+   * TestNG element was collected.
+   */
+  @BugPattern(severity = ERROR, summary = "Interacts with `TestNGScanner` for testing purposes")
+  public static final class TestChecker extends BugChecker implements CompilationUnitTreeMatcher {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public Description matchCompilationUnit(CompilationUnitTree tree, VisitorState state) {
+      TestNGScanner scanner = new TestNGScanner(state);
+      ImmutableMap<ClassTree, TestNGMetadata> classMetaData =
+          scanner.collectMetadataForEachClass(tree);
+
+      classMetaData.forEach(
+          (classTree, metaData) -> {
+            metaData
+                .getClassLevelAnnotationMetadata()
+                .ifPresent(
+                    annotation ->
+                        state.reportMatch(
+                            buildDescription(annotation.getAnnotationTree())
+                                .setMessage(
+                                    createAnnotationDiagnosticMessage(classTree, annotation, state))
+                                .build()));
+
+            metaData
+                .getDataProviderMetadata()
+                .values()
+                .forEach(
+                    dataProvider -> {
+                      state.reportMatch(
+                          buildDescription(dataProvider.getMethodTree())
+                              .setMessage(
+                                  String.format(
+                                      "class: %s dataProvider: %s",
+                                      classTree.getSimpleName(), dataProvider.getName()))
+                              .build());
+                    });
+
+            metaData
+                .getSetupMethods()
+                .forEach(
+                    (method, setupMethodType) ->
+                        state.reportMatch(
+                            buildDescription(method)
+                                .setMessage(
+                                    String.format(
+                                        "class: %s setupMethod: %s",
+                                        classTree.getSimpleName(), setupMethodType))
+                                .build()));
+
+            classTree.getMembers().stream()
```

Comment:
Should we move this to a `matchMethodTree` method?

Reply:
Have some thoughts about the setup of this class, maybe we can improve and make it cleaner :).